docs/user-guide/02-convenience-features.md

```diff
@@ -0,0 +1,227 @@
+
+## Member functions
+
+Slang supports defining member functions in `struct`s. For example, it is allowed to write:
+
+```hlsl
+struct Foo
+{
+    int compute(int a, int b)
+    {
+        return a + b;
+    }
+}
+```
+
+You can use the `.` syntax to invoke member functions:
+
+```hlsl
+Foo foo;
+int rs = foo.compute(1,2);
+```
+
+Slang also supports static member functions, For example:
+```
+struct Foo
+{
+    static int staticMethod(int a, int b)
+    {
+        return a + b;
+    }
+}
+```
+
+Static member functions are accessed the same way as other static members, via either the type name or an instance of the type:
+
+```hlsl
+int rs = Foo.staticMethod(a, b);
+```
+
+or
+
+```hlsl
+Foo foo;
+...
+int rs = foo.staticMethod(a,b);
+```
+
+### Mutability of member function
+
+For GPU performance considerations, the `this` argument in a member function is immutable by default. If you modify the content in `this` argument, the modification will be discarded after the call and does not affect the input object. If you intend to define a member function that mutates the object, use `[mutating]` attribute on the member function as shown in the following example.
+
+```hlsl
+struct Foo
+{
+    int count;
+    
+    [mutating]
+    void setCount(int x) { count = x; }
+
+    void setCount2(int x) { count = x; }
+}
+
+void test()
+{
+    Foo f;
+    f.setCount(1); // f.count is 1 after the call.
+    f.setCount2(2); // f.count is still 1 after the call.
+}
+```
+
+## Properties
+
+Properties provide a convienient way to access values exposed by a type, where the logic behind accessing the value is defined in `getter` and `setter` function pairs. Slang's `property` feature is similar to C# and Swift. 
+```C#
+struct MyType
+{
+    uint flag;
+
+    property uint highBits
+    {
+        get { return flag >> 16; }
+        set { flag = (flag & 0xFF) + (newValue << 16); }
+    }
+};
+```
+
+Or equivalently in a "modern" syntax:
+
+```C#
+struct MyType
+{
+    uint flag;
+
+    property highBits : uint
+    {
+        get { return flag >> 16; }
+        set { flag = (flag & 0xFF) + (newValue << 16); }
+    }
+};
+```
+
+You may also use an explicit parameter for the setter method:
+```C#
+property uint highBits
+{
+    set(uint x) { flag = (flag & 0xFF) + (x << 16);  }
+}
+```
+
+> #### Note ####
+> Slang currently does not support automatically synthesized `getter` and `setter` methods. For example,
+> the following code is not supported:
+> ```
+> property uint highBits {get;set;} // Not supported yet.
+> ```
+
+## Initializers
+> #### Note ####
+> The syntax for defining initializers are subject to future change.
+
+
+Slang supports defining initializers in `struct` types. You can write:
+```C#
+struct MyType
+{
+    int myVal;
+    __init(int a, int b)
+    {
+        myVal = a + b;
+    }
+}
+```
+
+You can use an initializer to construct a new instance by using the type name in a function call expression:
```

Comment:
We should probably refer to this form as just a *call expression* and not a *function call expression*, because of the fact that clearly one can call things that are not functions.